### PR TITLE
Enforce mypy compliance in CI

### PR DESCRIPTION
**Pull request by crusaderky**

### CODE REVIEW

**Comment by crusaderky** [comment]
mypy failures now make CI fail: https://dev.azure.com/xarray/xarray/_build/results?buildId=511&view=logs

Now moving to solve them...

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -408,17 +408,9 @@ def __init__(
...
-                "DataArray",
-                Variable,
-                Tuple[Hashable, Any],
-                Tuple[Sequence[Hashable], Any],
-            ],
-        ] = None,
+        data_vars: Mapping[Hashable, Any] = None,
         coords: Mapping[Hashable, Any] = None,
-        attrs: Mapping = None,
+        attrs: Mapping[Hashable, Any] = None,
```
Comment (crusaderky) [inline xarray/core/dataset.py]: Support ``Dataset({"x": [1, 2]})``, where one of the data_vars is silently moved to coords

**Comment by crusaderky** [comment]
Ready for review and merge

File: doc/whats-new.rst [hunk 2]
```diff
@@ -29,8 +29,8 @@ New functions/methods
   <http://github.com/DavidMertz>`_.
 
 - The xarray package is now discoverably by mypy (although typing hints
-  coverage is not complete yet). mypy users can now remove from their setup.cfg
-  the lines::
+  coverage is not complete yet). mypy type checking is now enforced by CI.
+  mypy users can now remove from their setup.cfg the lines::
```
Comment (max-sixty) [inline doc/whats-new.rst]:
```suggestion
  Libraries that depend xarray and use mypy can now remove from their setup.cfg the lines::
```
(or some variant - emphasizing that it's libraries?)

**Comment by max-sixty** [comment]
Is there a pre-commit for mypy?

**Comment by crusaderky** [comment]
@max-sixty added and tested :)

**Comment by shoyer** [comment]
Okay, we can give this a try

**Comment by crusaderky** [comment]
In it goes

**Pull request merged**